examples/chatbot.py

```diff
@@ -0,0 +1,191 @@
+os.environ['SSL_CERT_FILE'] = certifi.where()
+
+# Download UBER dataset
+os.system("mkdir -p data")
+os.system('wget "https://www.dropbox.com/s/948jr9cfs7fgj99/UBER.zip?dl=1" -O data/UBER.zip')
+os.system("unzip -o data/UBER.zip -d data")
+nltk.download('averaged_perceptron_tagger')
+
+# Load UBER documents
+years = [2022, 2021, 2020, 2019]
+loader = UnstructuredReader()
+doc_set = {}
+all_docs = []
+index_set = {}
+
+for year in years:
+    year_docs = loader.load_data(file=Path(f"./data/UBER/UBER_{year}.html"), split_documents=False)
+    # Insert year metadata into each document
+    for d in year_docs:
+        d.metadata = {"year": year}
+    doc_set[year] = year_docs
+    all_docs.extend(year_docs)
+
+Settings.chunk_size = 512
+Settings.chunk_overlap = 64
+Settings.llm = OpenAI(model="gpt-3.5-turbo")
+Settings.embed_model = OpenAIEmbedding(model="text-embedding-3-small")
+
+# Initialize simple vector indices
+index_set = {}
+for year in years:
+    storage_context = StorageContext.from_defaults()
+    cur_index = VectorStoreIndex.from_documents(
+        doc_set[year],
+        storage_context=storage_context,
+    )
+    index_set[year] = cur_index
+    storage_context.persist(persist_dir=f"./storage/{year}")
+
+individual_query_engine_tools = [
+    QueryEngineTool(
+        query_engine=index_set[year].as_query_engine(),
+        metadata=ToolMetadata(
+            name=f"vector_index_{year}",
+            description=f"useful for when you want to answer queries about the {year} SEC 10-K for Uber",
+        ),
+    )
+    for year in years
+]
+
+# Initialize SubQuestionQueryEngine
+query_engine = SubQuestionQueryEngine.from_defaults(
+    query_engine_tools=individual_query_engine_tools,)
+
+query_engine_tool = QueryEngineTool(
+    query_engine=query_engine,
+    metadata=ToolMetadata(
+        name="sub_question_query_engine",
+        description="useful for when you want to answer queries that require analyzing multiple SEC 10-K documents for Uber",
+    ),
+)
+tools = individual_query_engine_tools + [query_engine_tool]
+agent = OpenAIAgent.from_tools(tools, verbose=True)
+
+def extract_instructions(system_prompt):
+    instructions = system_prompt.split("\n")
+    instructions = [instruction.strip() for instruction in instructions if instruction.strip()]
+    return instructions
+
+
+def validate_and_format_json(data):
+    try:
+        json_string = json.dumps(data)
+        return json.loads(json_string)
+    except json.JSONDecodeError as e:
+        logging.error(f"Error parsing JSON: {e}")
+        raise
+
+def chatbot(user_query, instructions, openai_api_key, api_key):
+    openai.api_key = openai_api_key
+    contexts = []
+    relevance_scores = []
+    input_text = f"Instructions: {instructions}\nQuery: {user_query}"
+
+    logging.info("Sending request to OpenAI API...")
+    chat_response = agent.chat(input_text)
+    logging.info("Received response from OpenAI API.")
+
+    if hasattr(chat_response, 'source_nodes'):
```

Comment:
lets refactor line 118 through 129 into a function. Maybe call it `get_source_docs`

Reply:
done